File: app/src/main/java/com/mehmetakiftutuncu/muezzin/fragments/StatefulFragment.java
package com.mehmetakiftutuncu.muezzin.fragments;

import androidx.fragment.app.Fragment;

import com.kennyc.view.MultiStateView;

/**
 * Created by akif on 04/06/16.
 */
public abstract class StatefulFragment extends Fragment {
    protected MultiStateView multiStateViewLayout;

    public static final int RETRY_ACTION_DOWNLOAD = 1;

    abstract protected void changeStateTo(int newState, final int retryAction);
    abstract protected void retry(int action);
}

File: app/src/main/java/com/mehmetakiftutuncu/muezzin/fragments/StatefulFragment.kt
package com.mehmetakiftutuncu.muezzin.fragments

import android.content.Context
import android.os.Handler
import android.os.Looper
import androidx.fragment.app.Fragment
import com.kennyc.view.MultiStateView

abstract class StatefulFragment: Fragment() {
    protected lateinit var multiStateView: MultiStateView

    protected val ctx: Context by lazy { requireActivity() }

    protected val handler = Handler(Looper.getMainLooper())

    abstract fun changeStateTo(newState: Int, retryAction: Int)

    abstract fun retry(action: Int)

    protected inline fun runOnUI(crossinline f: () -> Unit) {
        requireActivity().runOnUiThread { f.invoke() }
    }

    companion object {
        const val retryActionDownload: Int = 1
    }
}